control/grpc.py

```diff
@@ -1442,6 +1455,62 @@ def namespace_resize(self, request, context=None):
 
         return pb2.req_status(status=ret.status, error_message=errmsg)
 
+    def namespace_recycle_safe(self, ana_id) ->int:
+        """Recycle namespaces."""
+        now = time.time()
+        self.logger.info(f"== recycle_safe started == for  anagrp{ana_id} time {now} ")
+
+        self.logger.info(f"Doing loop on {ana_id}  map; subsystem_nsid_anagrp:")
+        list_ns_params = []
+        for subsys, inner_dict in self.subsystem_nsid_anagrp.items():
+            for  ns_key, ana_value in inner_dict.items():
+                self.logger.info(f"nsid: {ns_key} ana_val: {ana_value}")
+                if ana_value == ana_id :
+                    self.logger.info(f"nsid {ns_key} for nqn {subsys} to recycle:")
+                    nsid = ns_key
+                    bdev_name = self.subsystem_nsid_bdev[subsys][nsid]
+                    assert  bdev_name
+                    ns_params = {'nsid':nsid, 'bdev_name':bdev_name, 'subsys':subsys}
+                    list_ns_params.append(ns_params)
+                    self.logger.info(f"nsid :{nsid}, pool_name: {self.bdev_params[bdev_name]['pool_name']}, rbd_name: {self.bdev_params[bdev_name]['image_name']}, block_size: {self.bdev_params[bdev_name]['block_size']}, uuid:{self.bdev_params[bdev_name]['uuid']}, anagrpid:{ana_id}")
+
+                    ret = self.remove_namespace(subsys, nsid, None)
+                    if ret.status != 0:
+                        return -1
+                    if bdev_name:
+                        ret_del = self.delete_bdev(bdev_name, True)
+                        if ret_del.status != 0:
+                            errmsg = f"Failure deleting namespace {nsid} from {subsys}: {ret_del.error_message}"
+                            self.logger.error(errmsg)
+                            return -1
```

Comment:
bdev_name is asserted at line 1472
```suggestion

                    ret_del = self.delete_bdev(bdev_name, True)
                    if ret_del.status != 0:
                        errmsg = f"Failure deleting namespace {nsid} from {subsys}: {ret_del.error_message}"
                        self.logger.error(errmsg)
                        return -1
```